src/build_context.rs

```diff
@@ -1132,6 +1148,19 @@ impl BuildContext {
         }
         Ok(wheels)
     }
+    /// Check if user requires to install rust toolchain
+    ///
+    /// Loop over `build-system.requires` defined in pyproject.toml and see if `rust-toolchain` is provided
```

Comment:
This isn't going to work because `pip` will try to install `rust-toolchain` package from `pypi`.

Reply:
Since we don't have `rust-toolchain` on pypi yet, we can try to use `rust-toolchain.toml` / legacy `rust-toolchain` file or `rust-version` in `Cargo.toml` of the binding crate, then fallback to Rust `stable`.

Reply:
So it means like if user provides a `rust-toolchain.toml` file in the project directory, that would be indicating that user wants to install rust-toolchain ?

Reply:
To put my two cents in, I think that the rust toolchain should be present for the build, even if `rust-toolchain.toml`is missing in which case the default toolchain would be used. Bonus points if `[toolchain]` can be specified directly in the `pyproject.toml` as e.g `[tool.maturin.toolchain]`.

Reply:
how about an additional command line flag such as `maturin sdist --install-rust-toolchain` ? (Or maybe when env variable `MATURIN_INSTALL_RUST` is set to true? ). This will be backward-compatible while keeping maturin flexible for users .

Reply:
This would only be suitable in the case the user of the package knows what Maturin, Rust etc. is. However, the reason I brought up issue #2120, is that a typical user just wants to use a package and would probably do not like to think about the build backend of this package.

In other words, it would be nice in my opinion, if a user can install a source distribution of a package that uses `maturin` as `build-system` simply via `pip install` without knowing such internals

Reply:
I found that currently maturin did check for the presence of rust toolchain when it attempts to build a project : 

https://github.com/PyO3/maturin/blob/main/maturin/__init__.py#L178-L182

So as an attempt to replicate this error msg, I tried to install a pypi package in an isolated environment without rust: 

`pip install --no-binary orjson orjson`

And I did see this error msg displayed in stdout. So maybe instead of exiting, we can attempt to install the toolchain from there ?

Reply:
@messense @padix-key I've modified `prepare_metadata_for_build_wheel` to remove the check for rust-toolchain, and add logic in `maturin pep517 write-dist-info` command to check for toolchain from there ( and install from rustup if not present). Let me know what you think

Reply:
I have actually barely insight into the `maturin` internals, I only know the user side 😃.

Reply:
I think automatically downloading an installation script and running it for the user and installing to non-standard locations is undesirable. I would not be happy if this took place on my system without any confirmation at least. I could also see it causing problems for CI systems which may accidentally install rustup each build if misconfigured.
An opt-in command line flag seems like a good approach to me. If not passed it could print
```
rust toolchain (cargo) not found.
Please install it manually or pass `maturin sdist --auto-install-toolchain`
```
so discoverability wouldn't be an issue for the 'typical user' described as not being aware of what they need to have installed